README.md
# Express Rate Limit

[![Tests](https://github.com/nfriedly/express-rate-limit/workflows/Test/badge.svg)](https://github.com/nfriedly/express-rate-limit/actions)
[![npm version](https://img.shields.io/npm/v/express-rate-limit.svg)](https://npmjs.org/package/express-rate-limit 'View this project on NPM')
[![npm downloads](https://img.shields.io/npm/dm/express-rate-limit)](https://www.npmjs.com/package/express-rate-limit)

Basic rate-limiting middleware for Express. Use to limit repeated requests to
public APIs and/or endpoints such as password reset.

Plays nice with
[express-slow-down](https://www.npmjs.com/package/express-slow-down).

> Note: this module does not share state with other processes/servers by
> default. If you need a more robust solution, I recommend using an external
> store. See the [`stores` section](#store) below for a list of external stores.

### Alternate Rate-limiters

This module was designed to only handle the basics and didn't even support
external stores initially. These other options all are excellent pieces of
software and may be more appropriate for some situations:

- [rate-limiter-flexible](https://www.npmjs.com/package/rate-limiter-flexible)
- [express-brute](https://www.npmjs.com/package/express-brute)
- [rate-limiter](https://www.npmjs.com/package/express-limiter)

## Install

From the npm registry:

```sh
# Using npm
> npm install express-rate-limit
# Using yarn or pnpm
> yarn/pnpm add express-rate-limit
```

From Github Releases:

```sh
# Using npm
> npm install https://github.com/nfriedly/express-rate-limit/releases/download/v{version}/express-rate-limit.tgz
# Using yarn or pnpm
> yarn/pnpm add https://github.com/nfriedly/express-rate-limit/releases/download/v{version}/express-rate-limit.tgz
```

Replace `{version}` with the version of the package that you want to your, e.g.:
`6.0.0`.

## Usage

This library is provided in ESM as well as CJS forms. To import it in a CJS
project:

```ts
const rateLimit = require('express-rate-limit')
```

To import it in a Typescript/ESM project:

```ts
import rateLimit from 'express-rate-limit'
```

### Examples

To use it in an API-only server where the rate-limiter should be applied to all
requests:

```ts
import rateLimit from 'express-rate-limit'

// Enable if you're behind a reverse proxy (Heroku, Bluemix, AWS ELB, Nginx, etc)
// see https://expressjs.com/en/guide/behind-proxies.html
// app.set('trust proxy', 1);

const limiter = rateLimit({
	windowMs: 15 * 60 * 1000, // 15 minutes
	max: 100, // Limit each IP to 100 requests per `window` (here, per 15 minutes)
	standardHeaders: true, // Return rate limit info in the `RateLimit-*` headers
	legacyHeaders: false, // Disable the `X-RateLimit-*` headers
})

// Apply the rate limiting middleware to all requests
app.use(limiter)
```

To use it in a 'regular' web server (e.g. anything that uses
`express.static()`), where the rate-limiter should only apply to certain
requests:

```ts
import rateLimit from 'express-rate-limit'

// Enable if you're behind a reverse proxy (Heroku, Bluemix, AWS ELB, Nginx, etc)
// see https://expressjs.com/en/guide/behind-proxies.html
// app.set('trust proxy', 1);

const apiLimiter = rateLimit({
	windowMs: 15 * 60 * 1000, // 15 minutes
	max: 100, // Limit each IP to 100 requests per `window` (here, per 15 minutes)
	standardHeaders: true, // Return rate limit info in the `RateLimit-*` headers
	legacyHeaders: false, // Disable the `X-RateLimit-*` headers
})

// Apply the rate limiting middleware to API calls only
app.use('/api', apiLimiter)
```

To create multiple instances to apply different rules to different endpoints:

```ts
import rateLimit from 'express-rate-limit'

// Enable if you're behind a reverse proxy (Heroku, Bluemix, AWS ELB, Nginx, etc)
// see https://expressjs.com/en/guide/behind-proxies.html
// app.set('trust proxy', 1);

const apiLimiter = rateLimit({
	windowMs: 15 * 60 * 1000, // 15 minutes
	max: 100, // Limit each IP to 100 requests per `window` (here, per 15 minutes)
	standardHeaders: true, // Return rate limit info in the `RateLimit-*` headers
	legacyHeaders: false, // Disable the `X-RateLimit-*` headers
})

app.use('/api/', apiLimiter)

const createAccountLimiter = rateLimit({
	windowMs: 60 * 60 * 1000, // 1 hour
	max: 5, // Limit each IP to 5 create account requests per `window` (here, per hour)
	message:
		'Too many accounts created from this IP, please try again after an hour',
	standardHeaders: true, // Return rate limit info in the `RateLimit-*` headers
	legacyHeaders: false, // Disable the `X-RateLimit-*` headers
})

app.post('/create-account', createAccountLimiter, (request, response) => {
	//...
})
```

To use a custom store:

```ts
import rateLimit from 'express-rate-limit'
import MemoryStore from 'express-rate-limit/memory-store.js'

// Enable if you're behind a reverse proxy (Heroku, Bluemix, AWS ELB, Nginx, etc)
// see https://expressjs.com/en/guide/behind-proxies.html
// app.set('trust proxy', 1);

const apiLimiter = rateLimit({
	windowMs: 15 * 60 * 1000, // 15 minutes
	max: 100, // Limit each IP to 100 requests per `window` (here, per 15 minutes)
	standardHeaders: true, // Return rate limit info in the `RateLimit-*` headers
	store: new MemoryStore(),
})

// Apply the rate limiting middleware to API calls only
app.use('/api', apiLimiter)
```

> **Note:** most stores will require additional configuration, such as custom
> prefixes, when using multiple instances. The default built-in memory store is
> an exception to this rule.

## Request API

A `request.rateLimit` property is added to all requests with the `limit`,
`current`, and `remaining` number of requests and, if the store provides it, a
`resetTime` Date object. These may be used in your application code to take
additional actions or inform the user of their status.

The property name can be configured with the configuration option
`requestPropertyName`

## Configuration options

### `windowMs`

Time frame for which requests are checked/remembered. Also used in the
`Retry-After` header when the limit is reached.

Note: with non-default stores, you may need to configure this value twice, once
here and once on the store. In some cases the units also differ (e.g. seconds vs
miliseconds)

Defaults to `60000` ms (= 1 minute).

### `max`

Max number of connections during `windowMs` milliseconds before sending a 429
response.

May be a number, or a function that returns a number or a promise. If `max` is a
function, it will be called with `request` and `response` params.

Defaults to `5`. Set to `0` to disable.

Example of using a function:

```ts
import rateLimit from 'express-rate-limit'

const isPremium = (request) => {
	// ...
}

const limiter = rateLimit({
	// `max` could also be an async function or return a promise
	max: (request, response) => {
		if (isPremium(request)) return 10
		else return 5
	},
	// ...
})

// Apply the rate limiting middleware to all requests
app.use(limiter)
```

### `message`

Error message sent to user when `max` is exceeded.

May be a `string`, JSON object, or any other value that Express's
[response.send](https://expressjs.com/en/4x/api.html#response.send) method
supports.

Defaults to `'Too many requests, please try again later.'`

### `statusCode`

HTTP status code returned when `max` is exceeded.

Defaults to `429`.

### `legacyHeaders`

Enable headers for request limit (`X-RateLimit-Limit`) and current usage
(`X-RateLimit-Remaining`) on all responses and time to wait before retrying
(`Retry-After`) when `max` is exceeded.

Defaults to `true`.

> Renamed in `6.x` from `headers` to `legacyHeaders`.

### `standardHeaders`

Enable headers conforming to the
[ratelimit standardization draft](https://github.com/ietf-wg-httpapi/ratelimit-headers/blob/main/draft-ietf-httpapi-ratelimit-headers.md)
adopted by the IETF: `RateLimit-Limit`, `RateLimit-Remaining`, and, if the store
supports it, `RateLimit-Reset`. May be used in conjunction with, or instead of
the `legacyHeaders` option.

This setting also enables the `Retry-After` header when `max` is exceeded.

Defaults to `false` (for backward compatibility), but recommended to use.

> Renamed in `6.x` from `draft_polli_ratelimit_headers` to `standardHeaders`.

### `keyGenerator`

Function used to generate keys.

Defaults to `request.ip`, similar to this:

```ts
const keyGenerator = (request /*, response*/) => request.ip
```

### `handler`

The function to handle requests once the max limit is exceeded. It receives the
`request` and the `response` objects. The `next` param is available if you need
to pass to the next middleware/route. Finally, the `optionsUsed` param has all
of the options that originally passed in when creating the current limiter and
the default values for other options.

The `request.rateLimit` object has `limit`, `current`, and `remaining` number of
requests and, if the store provides it, a `resetTime` Date object.

Defaults to:

```ts
const handler = (request, response, next, optionsUsed) => {
	response.status(options.statusCode).send(options.message)
}
```

### `onLimitReached`

Function that is called the first time a user hits the rate limit within a given
window.

The `request.rateLimit` object has the rate limit `limit`, number of requests
made so far `current`, the remaining number of requests `remaining` and, if the
store provides it, a Date object for when the hit count for all clients will be
reset `resetTime`.

Default is an empty function:

```ts
const onLimitReached = (request, response, optionsUsed) => {}
```

### `requestWasSuccessful`

Function that is called when `skipFailedRequests` and/or
`skipSuccessfulRequests` are set to `true`. May be overridden if, for example, a
service sends out a 200 status code on errors.

Defaults to

```ts
const requestWasSuccessful = (request, response) => response.statusCode < 400
```

### `skipFailedRequests`

When set to `true`, failed requests won't be counted. Request considered failed
when:

- response status >= 400
- requests that were cancelled before last chunk of data was sent (response
  `close` event triggered)
- response `error` event was triggrered by response

(Technically they are counted and then un-counted, so a large number of slow
requests all at once could still trigger a rate-limit. This may be fixed in a
future release.)

Defaults to `false`.

### `skipSuccessfulRequests`

When set to `true` successful requests (response status < 400) won't be counted.
(Technically they are counted and then un-counted, so a large number of slow
requests all at once could still trigger a rate-limit. This may be fixed in a
future release.)

Defaults to `false`.

### `skip`

Function used to skip (whitelist) requests. Returning `true`, or a promise that
resolves with `true`, from the function will skip limiting for that request.

Defaults to always `false` (count all requests):

```ts
const skip = (/*request, response*/) => false
```

### `requestPropertyName`

The name of the property that contains the rate limit information to add to the
`request` object.

Defaults to `rateLimit`.

### `store`

The storage to use when persisting rate limit attempts.

By default, the [memory store](source/memory-store.ts) is used.

Available data stores are:

- [memory-store](source/memory-store.ts): _(default)_ Simple in-memory option.
  Does not share state when app has multiple processes or servers.
- [rate-limit-redis](https://npmjs.com/package/rate-limit-redis): A
  [Redis](http://redis.io/)-backed store, more suitable for large or demanding
  deployments.
- [rate-limit-memcached](https://npmjs.org/package/rate-limit-memcached): A
  [Memcached](https://memcached.org/)-backed store.
- [rate-limit-mongo](https://www.npmjs.com/package/rate-limit-mongo): A
  [MongoDB](https://www.mongodb.com/)-backed store.
- [precise-memory-rate-limit](https://www.npmjs.com/package/precise-memory-rate-limit) -
  A memory store similar to the built-in one, except that it stores a distinct
  timestamp for each IP rather than bucketing them together.

You may also create your own store. It must implement the `Store` interface as
follows:

```ts
import rateLimit, {
	Store,
	Options,
	IncrementResponse,
} from 'express-rate-limit'

/**
 * A {@link Store} that stores the hit count for each client.
 *
 * @public
 */
class SomeStore implements Store {
	/**
	 * Some store-specific parameter.
	 */
	customParam!: string
	/**
	 * The duration of time before which all hit counts are reset (in milliseconds).
	 */
	windowMs!: number

	/**
	 * @constructor for {@link SomeStore}. Only required if the user needs to pass
	 * some store specific parameters. For example, in a Mongo Store, the user will
	 * need to pass the URI, username and password for the Mongo database.
	 *
	 * @param customParam {string} - Some store-specific parameter.
	 */
	constructor(customParam: string) {
		this.customParam = customParam
	}

	/**
	 * Method that actually initializes the store. Must be synchronous.
	 *
	 * @param options {Options} - The options used to setup the middleware.
	 *
	 * @public
	 */
	init(options: Options): void {
		this.windowMs = options.windowMs

		// ...
	}

	/**
	 * Method to increment a client's hit counter.
	 *
	 * @param key {string} - The identifier for a client
	 *
	 * @returns {IncrementResponse} - The number of hits and reset time for that client
	 *
	 * @public
	 */
	async increment(key: string): Promise<IncrementResponse> {
		// ...

		return {
			totalHits,
			resetTime,
		}
	}

	/**
	 * Method to decrement a client's hit counter.
	 *
	 * @param key {string} - The identifier for a client
	 *
	 * @public
	 */
	async decrement(key: string): Promise<void> {
		// ...
	}

	/**
	 * Method to reset a client's hit counter.
	 *
	 * @param key {string} - The identifier for a client
	 *
	 * @public
	 */
	async resetKey(key: string): Promise<void> {
		// ...
	}

	/**
	 * Method to reset everyone's hit counter.
	 *
	 * @public
	 */
	async resetAll(): Promise<void> {
		// ...
	}
}

export default SomeStore
```

## Instance API

### `resetKey(key)`

Resets the rate limiting for a given key. An example use case is to allow users
to complete a captcha or whatever to reset their rate limit, then call this
method.

## License

MIT © [Nathan Friedly](http://nfriedly.com/)


changelog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [7.1.4](https://github.com/express-rate-limit/express-rate-limit/releases/tag/v7.1.4)

### Fixed

- Ensure header values are strings rather than numbers, for compatibility with
  [Bun](https://bun.sh/)

## [7.1.3](https://github.com/express-rate-limit/express-rate-limit/releases/tag/v7.1.3)

### Changed

- Loosened peer dependencies to explicitly allow the Express 5 beta. (See
  [#415](https://github.com/express-rate-limit/express-rate-limit/issues/415))

## [7.1.2](https://github.com/express-rate-limit/express-rate-limit/releases/tag/v7.1.2)

### Changed

- Re-organized documentation from readme into docs/ folder and added
  documentation website.

## [v7.1.1](https://github.com/express-rate-limit/express-rate-limit/releases/tag/v7.1.1)

### Added

- Enabled provenance statement generation, see
  https://github.com/express-rate-limit/express-rate-limit#406.

## [7.1.0](https://github.com/express-rate-limit/express-rate-limit/releases/tag/v7.1.0)

### Changed

- The `getKey` method is now always defined. If the store does not have the
  required `get` method, `getKey` will throw an error explaining this.

## [7.0.2](https://github.com/express-rate-limit/express-rate-limit/releases/tag/v7.0.2)

### Added

- Added `cluster-memory-store` to the readme and made a couple of other minor
  clarifications.

## [7.0.1](https://github.com/express-rate-limit/express-rate-limit/releases/tag/v7.0.1)

### Added

- Added `rate-limit-postgresql` to the `stores` list in the readme.

## [7.0.0](https://github.com/express-rate-limit/express-rate-limit/releases/tag/v7.0.0)

### Breaking

- Changed behavior when `max` is set to 0:
  - Previously, `max: 0` was treated as a 'disable' flag and would allow all
    requests through.
  - Starting with v7, all requests will be blocked when max is set to 0.
  - To replicate the old behavior, use the
    [skip](https://github.com/express-rate-limit/express-rate-limit#skip)
    function instead.
- Renamed `req.rateLimit.current` to `req.rateLimit.used`.
  - `current` is now a hidden getter that will return the `used` value, but it
    will not appear when iterating over the keys or calling `JSON.stringify()`.
- Changed the minimum required Node version from v14 to v16.
  - `express-rate-limit` now targets `es2022` in TypeScript/ESBuild.
- Bumped TypeScript from v4 to v5 and `dts-bundle-generator` from v7 to v8.

### Deprecated

- Removed the `draft_polli_ratelimit_headers` option (it was deprecated in v6).
  - Use `standardHeaders: 'draft-6'` instead.
- Removed the `onLimitReached` option (it was deprecated in v6).
  - [This](<(https://github.com/express-rate-limit/express-rate-limit/wiki/Error-Codes#wrn_erl_deprecated_on_limit_reached)>)
    is an example of how to replicate it's behavior with a custom `handler`
    option.

### Changed

- The `MemoryStore` now uses precise, per-user reset times rather than a global
  window that resets all users at once.
- The `limit` configuration option is now prefered to `max`.
  - It still shows the same behavior, and `max` is still supported. The change
    was made to better align with terminology used in the IETF standard drafts.

### Added

- The `validate` config option can now be an object with keys to enable or
  disable specific validation checks. For more information, see
  [this](https://github.com/express-rate-limit/express-rate-limit#validate).

## [6.11.2](https://github.com/express-rate-limit/express-rate-limit/releases/tag/v6.11.2)

### Fixed

- Restored `IncrementResponse ` TypeScript type (See
  [#397](https://github.com/express-rate-limit/express-rate-limit/pull/397))

## [6.11.1](https://github.com/express-rate-limit/express-rate-limit/releases/tag/v6.11.1)

### Fixed

- Check for prefixed keys when validating that the stores have single counted
  keys (See
  [#395](https://github.com/express-rate-limit/express-rate-limit/issues/395)).

## [6.11.0](https://github.com/express-rate-limit/express-rate-limit/releases/tag/v6.11.0)

### Added

- Support for retrieving the current hit count and reset time for a given key
  from a store (See
  [#390](https://github.com/express-rate-limit/express-rate-limit/issues/389)).

## [6.10.0](https://github.com/express-rate-limit/express-rate-limit/releases/tag/v6.10.0)

### Added

- Support for combined `RateLimit` header from the
  [RateLimit header fields for HTTP standardization draft](https://github.com/ietf-wg-httpapi/ratelimit-headers)
  adopted by the IETF. Enable by setting `standardHeaders: 'draft-7'`.
- New `standardHeaders: 'draft-6'` option, treated equivalent to
  `standardHeaders: true` from previous releases. Note that `true` and `false`
  are still supported.
- New `RateLimit-Policy` header added when `standardHeaders` is set to
  `'draft-6'`, `'draft-7'`, or `true`.
- Warning when using deprecated `draft_polli_ratelimit_headers` option.
- Warning when using deprecated `onLimitReached` option.
- Warning when `totalHits` value returned from Store is invalid.

## [6.9.0](https://github.com/express-rate-limit/express-rate-limit/releases/tag/v6.9.0)

### Added

- New validaion check for double-counted requests.
- Added help link to each validation error, directing users to the appropriate
  wiki page for more info.

### Changed

- Miscellaneous documenation improvements.

## [6.8.1](https://github.com/express-rate-limit/express-rate-limit/releases/tag/v6.8.0) & [6.7.2](https://github.com/express-rate-limit/express-rate-limit/releases/tag/v6.8.0)

### Changed

- Revert 6.7.1 change that bumped typescript from 5.x to 4.x and
  dts-bundle-generator from 8.x to 7.x (See
  [#360](https://github.com/express-rate-limit/express-rate-limit/issues/360)).


source/lib.ts
// /source/lib.ts
// The option parser and rate limiting middleware

import type { Request, Response, NextFunction, RequestHandler } from 'express'
import type {
	Options,
	AugmentedRequest,
	RateLimitRequestHandler,
	LegacyStore,
	Store,
	ClientRateLimitInfo,
	ValueDeterminingMiddleware,
	RateLimitExceededEventHandler,
	DraftHeadersVersion,
	RateLimitInfo,
	EnabledValidations,
} from './types.js'
import {
	setLegacyHeaders,
	setDraft6Headers,
	setDraft7Headers,
	setRetryAfterHeader,
} from './headers.js'
import { getValidations, type Validations } from './validations.js'
import MemoryStore from './memory-store.js'

/**
 * Type guard to check if a store is legacy store.
 *
 * @param store {LegacyStore | Store} - The store to check.
 *
 * @return {boolean} - Whether the store is a legacy store.
 */
const isLegacyStore = (store: LegacyStore | Store): store is LegacyStore =>
	// Check that `incr` exists but `increment` does not - store authors might want
	// to keep both around for backwards compatibility.
	typeof (store as any).incr === 'function' &&
	typeof (store as any).increment !== 'function'

/**
 * Converts a legacy store to the promisified version.
 *
 * @param store {LegacyStore | Store} - The store passed to the middleware.
 *
 * @returns {Store} - The promisified version of the store.
 */
const promisifyStore = (passedStore: LegacyStore | Store): Store => {
	if (!isLegacyStore(passedStore)) {
		// It's not an old store, return as is
		return passedStore
	}

	const legacyStore = passedStore

	// A promisified version of the store
	class PromisifiedStore implements Store {
		async increment(key: string): Promise<ClientRateLimitInfo> {
			return new Promise((resolve, reject) => {
				legacyStore.incr(
					key,
					(
						error: Error | undefined,
						totalHits: number,
						resetTime: Date | undefined,
					) => {
						if (error) reject(error)
						resolve({ totalHits, resetTime })
					},
				)
			})
		}

		async decrement(key: string): Promise<void> {
			return legacyStore.decrement(key)
		}

		async resetKey(key: string): Promise<void> {
			return legacyStore.resetKey(key)
		}

		/* istanbul ignore next */
		async resetAll(): Promise<void> {
			if (typeof legacyStore.resetAll === 'function')
				return legacyStore.resetAll()
		}
	}

	return new PromisifiedStore()
}

/**
 * The internal configuration interface.
 *
 * This is copied from Options, with fields made non-readonly and deprecated
 * fields removed.
 *
 * For documentation on what each field does, {@see Options}.
 *
 * This is not stored in types because it's internal to the API, and should not
 * be interacted with by the user.
 */
type Configuration = {
	windowMs: number
	limit: number | ValueDeterminingMiddleware<number>
	message: any | ValueDeterminingMiddleware<any>
	statusCode: number
	legacyHeaders: boolean
	standardHeaders: false | DraftHeadersVersion
	requestPropertyName: string
	skipFailedRequests: boolean
	skipSuccessfulRequests: boolean
	keyGenerator: ValueDeterminingMiddleware<string>
	handler: RateLimitExceededEventHandler
	skip: ValueDeterminingMiddleware<boolean>
	requestWasSuccessful: ValueDeterminingMiddleware<boolean>
	store: Store
	validations: Validations
}

/**
 * Converts a `Configuration` object to a valid `Options` object, in case the
 * configuration needs to be passed back to the user.
 *
 * @param config {Configuration} - The configuration object to convert.
 *
 * @returns {Partial<Options>} - The options derived from the configuration.
 */
const getOptionsFromConfig = (config: Configuration): Options => {
	const { validations, ...directlyPassableEntries } = config

	return {
		...directlyPassableEntries,
		validate: validations.enabled as EnabledValidations,
	}
}

/**
 *
 * Remove any options where their value is set to undefined. This avoids overwriting defaults
 * in the case a user passes undefined instead of simply omitting the key.
 *
 * @param passedOptions {Options} - The options to omit.
 *
 * @returns {Options} - The same options, but with all undefined fields omitted.
 *
 * @private
 */
const omitUndefinedOptions = (
	passedOptions: Partial<Options>,
): Partial<Options> => {
	const omittedOptions: Partial<Options> = {}

	for (const k of Object.keys(passedOptions)) {
		const key = k as keyof Options

		if (passedOptions[key] !== undefined) {
			// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
			omittedOptions[key] = passedOptions[key]
		}
	}

	return omittedOptions
}

/**
 * Type-checks and adds the defaults for options the user has not specified.
 *
 * @param options {Options} - The options the user specifies.
 *
 * @returns {Configuration} - A complete configuration object.
 */
const parseOptions = (passedOptions: Partial<Options>): Configuration => {
	// Passing undefined should be equivalent to not passing an option at all, so we'll
	// omit all fields where their value is undefined.
	const notUndefinedOptions: Partial<Options> =
		omitUndefinedOptions(passedOptions)

	// Create the validator before even parsing the rest of the options.
	const validations = getValidations(notUndefinedOptions?.validate ?? true)
	validations.validationsConfig()

	// Warn for the deprecated options. Note that these options have been removed
	// from the type definitions in v7.
	validations.draftPolliHeaders(
		// @ts-expect-error see the note above.
		notUndefinedOptions.draft_polli_ratelimit_headers,
	)
	// @ts-expect-error see the note above.
	validations.onLimitReached(notUndefinedOptions.onLimitReached)

	// The default value for the `standardHeaders` option is `false`. If set to
	// `true`, it resolve to `draft-6`. `draft-7` (recommended) is used only if
	// explicitly set.
	let standardHeaders = notUndefinedOptions.standardHeaders ?? false
	if (standardHeaders === true) standardHeaders = 'draft-6'

	// See ./types.ts#Options for a detailed description of the options and their
	// defaults.
	const config: Configuration = {
		windowMs: 60 * 1000,
		limit: passedOptions.max ?? 5, // `max` is deprecated, but support it anyways.
		message: 'Too many requests, please try again later.',
		statusCode: 429,
		legacyHeaders: passedOptions.headers ?? true,
		requestPropertyName: 'rateLimit',
		skipFailedRequests: false,
		skipSuccessfulRequests: false,
		requestWasSuccessful: (_request: Request, response: Response): boolean =>
			response.statusCode < 400,
		skip: (_request: Request, _response: Response): boolean => false,
		keyGenerator(request: Request, _response: Response): string {
			// Run the validation checks on the IP and headers to make sure everything
			// is working as intended.
			validations.ip(request.ip)
			validations.trustProxy(request)
			validations.xForwardedForHeader(request)

			// By default, use the IP address to rate limit users.
			// eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
			return request.ip!
		},
		async handler(
			request: Request,
			response: Response,
			_next: NextFunction,
			_optionsUsed: Options,
		): Promise<void> {
			// Set the response status code.
			response.status(config.statusCode)
			// Call the `message` if it is a function.
			const message: unknown =
				typeof config.message === 'function'
					? await (config.message as ValueDeterminingMiddleware<any>)(
							request,
							response,
					  )
					: config.message

			// Send the response if writable.
			if (!response.writableEnded) {
				response.send(message)
			}
		},
		// Allow the default options to be overriden by the options passed to the middleware.
		...notUndefinedOptions,
		// `standardHeaders` is resolved into a draft version above, use that.
		standardHeaders,
		// Note that this field is declared after the user's options are spread in,
		// so that this field doesn't get overriden with an un-promisified store!
		store: promisifyStore(notUndefinedOptions.store ?? new MemoryStore()),
		// Print an error to the console if a few known misconfigurations are detected.
		validations,
	}

	// Ensure that the store passed implements the `Store` interface
	if (
		typeof config.store.increment !== 'function' ||
		typeof config.store.decrement !== 'function' ||
		typeof config.store.resetKey !== 'function' ||
		(config.store.resetAll !== undefined &&
			typeof config.store.resetAll !== 'function') ||
		(config.store.init !== undefined && typeof config.store.init !== 'function')
	) {
		throw new TypeError(
			'An invalid store was passed. Please ensure that the store is a class that implements the `Store` interface.',
		)
	}

	return config
}

/**
 * Just pass on any errors for the developer to handle, usually as a HTTP 500
 * Internal Server Error.
 *
 * @param fn {RequestHandler} - The request handler for which to handle errors.
 *
 * @returns {RequestHandler} - The request handler wrapped with a `.catch` clause.
 *
 * @private
 */
const handleAsyncErrors =
	(fn: RequestHandler): RequestHandler =>
	async (request: Request, response: Response, next: NextFunction) => {
		try {
			await Promise.resolve(fn(request, response, next)).catch(next)
		} catch (error: unknown) {
			/* istanbul ignore next */
			next(error)
		}
	}

/**
 *
 * Create an instance of IP rate-limiting middleware for Express.
 *
 * @param passedOptions {Options} - Options to configure the rate limiter.
 *
 * @returns {RateLimitRequestHandler} - The middleware that rate-limits clients based on your configuration.
 *
 * @public
 */
const rateLimit = (
	passedOptions?: Partial<Options>,
): RateLimitRequestHandler => {
	// Parse the options and add the default values for unspecified options
	const config = parseOptions(passedOptions ?? {})
	const options = getOptionsFromConfig(config)

	// Call the `init` method on the store, if it exists
	if (typeof config.store.init === 'function') config.store.init(options)

	// Then return the actual middleware
	const middleware = handleAsyncErrors(
		async (request: Request, response: Response, next: NextFunction) => {
			// First check if we should skip the request
			const skip = await config.skip(request, response)
			if (skip) {
				next()
				return
			}

			// Create an augmented request
			const augmentedRequest = request as AugmentedRequest

			// Get a unique key for the client
			const key = await config.keyGenerator(request, response)
			// Increment the client's hit counter by one.
			const { totalHits, resetTime } = await config.store.increment(key)
			// Make sure that -
			// - the hit count is incremented only by one.
			// - the returned hit count is a positive integer.
			config.validations.positiveHits(totalHits)
			config.validations.singleCount(request, config.store, key)

			// Get the limit (max number of hits) for each client.
			const retrieveLimit =
				typeof config.limit === 'function'
					? config.limit(request, response)
					: config.limit
			const limit = await retrieveLimit
			config.validations.limit(limit)

			// Define the rate limit info for the client.
			const info: RateLimitInfo = {
				limit,
				used: totalHits,
				remaining: Math.max(limit - totalHits, 0),
				resetTime,
			}

			// Set the `current` property on the object, but hide it from iteration
			// and `JSON.stringify`. See the `./types#RateLimitInfo` for details.
			Object.defineProperty(info, 'current', {
				configurable: false,
				enumerable: false,
				value: totalHits,
			})

			// Set the rate limit information on the augmented request object
			augmentedRequest[config.requestPropertyName] = info

			// Set the `X-RateLimit` headers on the response object if enabled.
			if (config.legacyHeaders && !response.headersSent) {
				setLegacyHeaders(response, info)
			}

			// Set the standardized `RateLimit-*` headers on the response object if
			// enabled.
			if (config.standardHeaders && !response.headersSent) {
				if (config.standardHeaders === 'draft-6') {
					setDraft6Headers(response, info, config.windowMs)
				} else if (config.standardHeaders === 'draft-7') {
					config.validations.headersResetTime(info.resetTime)
					setDraft7Headers(response, info, config.windowMs)
				}
			}

			// If we are to skip failed/successfull requests, decrement the
			// counter accordingly once we know the status code of the request
			if (config.skipFailedRequests || config.skipSuccessfulRequests) {
				let decremented = false
				const decrementKey = async () => {
					if (!decremented) {
						await config.store.decrement(key)
						decremented = true
					}
				}

				if (config.skipFailedRequests) {
					response.on('finish', async () => {
						if (!config.requestWasSuccessful(request, response))
							await decrementKey()
					})
					response.on('close', async () => {
						if (!response.writableEnded) await decrementKey()
					})
					response.on('error', async () => {
						await decrementKey()
					})
				}

				if (config.skipSuccessfulRequests) {
					response.on('finish', async () => {
						if (config.requestWasSuccessful(request, response))
							await decrementKey()
					})
				}
			}

			// Disable the validations, since they should have run at least once by now.
			config.validations.disable()

			// If the client has exceeded their rate limit, set the Retry-After header
			// and call the `handler` function.
			if (totalHits > limit) {
				if (config.legacyHeaders || config.standardHeaders) {
					setRetryAfterHeader(response, info, config.windowMs)
				}

				config.handler(request, response, next, options)
				return
			}

			next()
		},
	)

	const getThrowFn = () => {
		throw new Error('The current store does not support the get/getKey method')
	}

	// Export the store's function to reset and fetch the rate limit info for a
	// client based on their identifier.
	;(middleware as RateLimitRequestHandler).resetKey =
		config.store.resetKey.bind(config.store)
	;(middleware as RateLimitRequestHandler).getKey =
		typeof config.store.get === 'function'
			? config.store.get.bind(config.store)
			: getThrowFn

	return middleware as RateLimitRequestHandler
}

// Export it to the world!
export default rateLimit
